File: POO - Python & Java/practica-2/gestorAplicacion/Hospital.py
import random
from gestorAplicacion.Especialidad import Especialidad
from gestorAplicacion.Quirofano import Quirofano
from gestorAplicacion.Doctor import Doctor
from gestorAplicacion.Paciente import Paciente


class Hospital:

    #	CONSTRUCTORS	
    def __init__(self, salaDeEspera, quirofanos, doctores):
        
        self._denuncias = 0
        self._horario = None
        self._presupuesto = 0
        self._bioseguro = False

        self._salaDeEspera = salaDeEspera
        self._quirofanos = quirofanos
        self._doctores = doctores


    #	INSTANCE METHODS  

    #	
    #	 * despedirMedico: Busca la cedula del medico a despedir
    #	 * lo "indemniza"(resta del presupuesto del hospital el salario del medico)
    #	 
    def despedirDoctor(self, cedula, justificado):
        i = 0
        while i < len(self._doctores):
            if self._doctores[i] is not None and cedula == self._doctores[i].getCedula():
                if not justificado:
                    salario_doc = self._doctores[i].getEspecialidad()._salario
                    self._presupuesto -= (self._doctores[i].getEspecialidad().getSalario())
                self._doctores[i] = None
                break
            i += 1


    #	
    #	 *getDoctorQuirofano: hace un arreglo de los doctores que estan en los quirofanos
    #	 * es decir retorna un array de instancias de tipo doctores que estan en los quirofano, recuerde que los quirofanos 
    #	 * tambien son instacias el cual tiene un atributo referenciado de tipo doctor
    #	 
    def getDoctorQuirofano(self):
        quirofanos = self._quirofanos
        contador =0

        doctores = [None for _ in range(6)]
        i = 0
        while i < len(quirofanos):
            if quirofanos[i] is not None and quirofanos[i].getDoctor() is not None:
                doctores[contador]=quirofanos[i].getDoctor()
                contador += 1
            i += 1
        return doctores



    #	
    #	 * getMuestraDoctores: Retorna un array de 3 elementos de tipo Doctor
    #	 * Que pueden corresponder a una misma especialidad o no(sobrecarga)
    #	 
    def getMuestraDoctores(self):
        muestra = [None for _ in range(3)]

        i = 0
        for doctor in self._doctores:
            if doctor is not None:
                muestra[i] = doctor
                i += 1

        return muestra

    def getMuestraDoctores(self, especialidad):
        muestra = [None for _ in range(3)]

        i = 0
        for doctor in self._doctores:
            if doctor is not None and doctor.getEspecialidad() == especialidad:
                muestra[i] = doctor
                i += 1

        return muestra

    #	
    #	 * getNDoctoresEspecialidad: Retorna la cantidad de doctores que haya en dicha especialidad
    #	 
    def getNDoctoresEspecialidad(self, especialidad):
        i =0
        for doctor in self._doctores:
            if doctor is not None and doctor.getEspecialidad() == especialidad:
                i += 1
        return i

    #	
    #	 * implementarBioseguridad: implementa medidas de bioseguridad en el hospital
    #	 
    def implementarBioseguridad(self):

        # Guardando en un ArrayList una lista de las personas que se echaron
        personasHechadas = self.getSalaDeEspera().implementarBioseguridad()

        for persona in personasHechadas:
            if persona.isAgresivo():
                self._denuncias += 1

        # ITERANDO SOBRE CADA DOCTOR PARA VER SI ESTA DISPUESTO A TRABAJAR EN PANDEMIA
        for doctor in self.getDoctores():
            if doctor is not None and not doctor.isVoluntadPandemia():
                # EL DOCTOR "RENUNCIA"
                self.despedirDoctor(doctor.getCedula(), False)
                personasHechadas.append(doctor)
        self.setBioseguro(True)
        return personasHechadas

    #	
    #	 * apartarQuirofano(int index, Paciente paciente, Doctor doctor)
    #	 * y retorna un boolean para saber si se pudo apartar el quirofano
    #	 
    def apartarQuirofano(self, index, paciente, doctor):
        suceed = False
        quirofano = self.getQuirofanos()[index]
        if quirofano is None:
            self.getQuirofanos()[index] = Quirofano(paciente, doctor)
            suceed = True
        elif quirofano.getDoctor() is None:
            quirofano.setDoctor(doctor)
            quirofano.setPaciente(paciente)
            suceed = True
        return suceed


    #	
    #	 * apartarQuirofanoUrgencia: aparta el quirofano que encuentre disponible el paciente
    #	 * que este en parametro
    #	 * @params: Paciente paciente
    #	 
    def apartarQuirofanoUrgencia(self, paciente):
        # recorre la lista de quirofano verificando si tiene un paciente  
        cirujano = None
        quirofanos = self.getQuirofanos()
        i = 0
        while i < len(quirofanos):
            if quirofanos[i] is None or quirofanos[i].getDoctor() is None:

                # Busca por cada problema medico un cirujano disponible y asigna el primero
                for tipoProblema in paciente.getTipoProblemasMedicos():
                    cirujano = self.getcirujanoEspecialidad(tipoProblema)
                    if cirujano is not None:
                        break

                # Si no hay doctores disponibles, entonces conseguir un cirujano cualquiera
                #				if(cirujano == null) { 
                #					cirujano = this.getcirujanoEspecialidad(Especialidad.GENERAL); 
                #				}

                # Si no hay cirujanos disponibles, entonces conseguir un doctor general para que opere
                # for(Especialidad)

                # Si definitiva no encontro doctor, entonces F
                if cirujano is None:
                    return None

                # Si llego hasta aca es que si hay cirujano, por lo que creemos uno
                # para que ese elemento de la lista no apunte hacia un null sino hacia un Quirofano
                quirofanos[i] = Quirofano(paciente, cirujano)
                return quirofanos[i]
            i += 1
        # Si se retorno null es porque no hay quirofanos disponibles
        return None

    #	
    #	 * getDoctorEspecialidad(Especialidad especialidad)
    #	 
    def getcirujanoEspecialidad(self, especialidad):
        for doctor in self.getDoctores():
            if doctor is not None and doctor.getEspecialidad() == especialidad and doctor.isCirujano():
                return doctor
        return None

    #	
    #	 * esparcirVirus: se esparce el virus a los pacientes en la sala de espera
    #	 * puede ser el caso que el paciente se muera al ser poblacion de riesgo y por mala suerte(azar)
    #	 
    def esparcirVirus(self):
        personasHospital = self.getPersonasHospital()
        muertos = []

        for persona in personasHospital:
            if persona.isPoblacionDeRiesgo() and random.choice([True, False]):
                muerto = persona.infectar()
                if muerto:
                    muertos.append(persona)
                    if isinstance(persona, Paciente):
                        self.getSalaDeEspera().sacarPaciente(persona.getCedula())
                    elif isinstance(persona, Doctor):
                        self.despedirDoctor(persona.getCedula(), True)
        return muertos



    #	
    #	 * 
    #	 
    def getPersonasHospital(self):
        personasHospital = []

        for persona in self.getSalaDeEspera().getPacientes():
            if persona is not None:
                personasHospital.append(persona)

        for persona in self.getDoctores():
            if persona is not None:
                personasHospital.append(persona)
        return personasHospital

    #	GETTERS & SETTERS	
    def getHorario(self):
        return self.horario
    def delCita(self, cita):
        self.horario.remove(cita)
    def addCita(self, cita):
        self.horario.add(cita)

    def getDenuncias(self):
        return self._denuncias
    def setDenuncias(self, denuncias):
        self._denuncias = denuncias

    def getSalaDeEspera(self):
        return self._salaDeEspera
    def setSalaDeEspera(self, salaDeEspera):
        self._salaDeEspera = salaDeEspera

    def getQuirofanos(self):
        return self._quirofanos
    def setQuirofanos(self, quirofanos):
        self._quirofanos = quirofanos

    def getCuartos(self):
        return self._cuartos
    def setCuartos(self, cuartos):
        self._cuartos = cuartos

    def getDoctores(self):
        return self._doctores
    def setDoctores(self, doctores):
        self._doctores = doctores

    def getPresupuesto(self):
        return self._presupuesto
    def setPresupuesto(self, presupuesto):
        self._presupuesto = presupuesto

    def isBioseguro(self):
        return self._bioseguro
    def setBioseguro(self, bioseguro):
        self._bioseguro = bioseguro

    

File: POO - Python & Java/practica-1/src/gestorAplicacion/Hospital.java
package gestorAplicacion;

import java.util.Random;
import java.io.Serializable;	

import javax.swing.text.html.HTMLDocument.RunElement;

import java.util.ArrayList;

public class Hospital implements Serializable{
	
	/*	INSTANCE VARIABLES	*/
	private static final long serialVersionUID = 1L;
	private int denuncias = 0;
	private ArrayList<Cita> horario;
	private SalaDeEspera salaDeEspera;
	private Quirofano[] quirofanos;
	private Cuarto[] cuartos;
	private Doctor[] doctores;
	private int presupuesto;
	private boolean bioseguro;
	
	/*	CONSTRUCTORS	*/
	public Hospital(SalaDeEspera salaDeEspera, Quirofano[] quirofanos, Cuarto[] cuartos, Doctor[] doctores) {
		this.salaDeEspera = salaDeEspera;
		this.quirofanos = quirofanos;
		this.cuartos = cuartos;
		this.doctores = doctores;
	}
	
	
	/*	INSTANCE METHODS  */
	
	/*
	 * despedirMedico: Busca la cedula del medico a despedir
	 * lo "indemniza"(resta del presupuesto del hospital el salario del medico)
	 */
	public void despedirDoctor(int cedula,boolean justificado) {
		for (int i = 0; i < doctores.length; i++) {
			if(doctores[i] != null && cedula == doctores[i].getCedula()) {
				if(!justificado){
					this.presupuesto -= doctores[i].getEspecialidad().getSalario();
				}
				doctores[i] = null;
				break;
			}
		}
		
	}
		
	/*
	 *getDoctorQuirofano: hace un arreglo de los doctores que estan en los quirofanos
	 * es decir retorna un array de instancias de tipo doctores que estan en los quirofano, recuerde que los quirofanos 
	 * tambien son instacias el cual tiene un atributo referenciado de tipo doctor
	 */
	public Doctor[] getDoctorQuirofano(){
		Quirofano[] quirofanos= this.quirofanos;
		int contador=0;

		Doctor[] doctores= new Doctor[6];
		for (int i = 0; i < quirofanos.length; i++) {
            if(quirofanos[i]!=null){
				doctores[contador]=quirofanos[i].getDoctor();
                contador++;
            }
		}
		return doctores;


	}
	
	/*
	 * getMuestraDoctores: Retorna un array de 3 elementos de tipo Doctor
	 * Que pueden corresponder a una misma especialidad o no(sobrecarga)
	 */
	public Doctor[] getMuestraDoctores() {
		Doctor[] muestra = new Doctor[3];
		
		int i = 0;
		for (Doctor doctor : this.doctores) {
			if (doctor != null) {
				muestra[i++] = doctor;
			}
		}
		
		return muestra;
	}

	public Doctor[] getMuestraDoctores(Especialidad especialidad) {
		Doctor[] muestra = new Doctor[3];
		
		int i = 0;
		for (Doctor doctor : this.doctores) {
			if (doctor != null && doctor.getEspecialidad() == especialidad) {
				muestra[i++] = doctor;
			}
		}
		
		return muestra;
	}
	
	/*
	 * getNDoctoresEspecialidad: Retorna la cantidad de doctores que haya en dicha especialidad
	 */
	public int getNDoctoresEspecialidad(Especialidad especialidad) {
		int i=0;
		for (Doctor doctor : this.doctores) {
			if(doctor != null && doctor.getEspecialidad() == especialidad) {
				i++;
			}
		}
		return i;
	}
	
	/*
	 * implementarBioseguridad: implementa medidas de bioseguridad en el hospital
	 */
	public ArrayList<Persona> implementarBioseguridad() {
		
		// Guardando en un ArrayList una lista de las personas que se echaron
		ArrayList<Persona> personasHechadas = this.getSalaDeEspera().implementarBioseguridad();
		
		for (Persona persona : personasHechadas) {
			if (persona.isAgresivo()) {
				this.denuncias++;
			}
		}
		
		/* ITERANDO SOBRE CADA DOCTOR PARA VER SI ESTA DISPUESTO A TRABAJAR EN PANDEMIA*/
		for(Doctor doctor: this.getDoctores()) {
			if(doctor != null && !doctor.isVoluntadPandemia()) {
				// EL DOCTOR "RENUNCIA"
				this.despedirDoctor(doctor.getCedula(), false);
				personasHechadas.add(doctor);
			}
		}
		this.setBioseguro(true);
		return personasHechadas;
	}
	
	/*
	 * apartarQuirofano(int index, Paciente paciente, Doctor doctor)
	 * y retorna un boolean para saber si se pudo apartar el quirofano
	 */
	public boolean apartarQuirofano(int index, Paciente paciente, Doctor doctor) {
		boolean suceed = false;
		Quirofano quirofano = this.getQuirofanos()[index];
		if(quirofano == null) {
			this.getQuirofanos()[index] = new Quirofano(paciente, doctor);
			suceed = true;
		} else if (quirofano.getDoctor() == null) {
			quirofano.setDoctor(doctor);
			quirofano.setPaciente(paciente);
			suceed = true;
		}
		return suceed;
	}
	
	
	/*
	 * apartarQuirofanoUrgencia: aparta el quirofano que encuentre disponible el paciente
	 * que este en parametro
	 * @params: Paciente paciente
	 */
	public Quirofano apartarQuirofanoUrgencia(Paciente paciente){
		/* recorre la lista de quirofano verificando si tiene un paciente */ 
		Doctor cirujano = null;
		Quirofano[] quirofanos = this.getQuirofanos();
		for(int i = 0; i < quirofanos.length; i++){
			if (quirofanos[i]==null || quirofanos[i].getDoctor()==null){
				
				// Busca por cada problema medico un cirujano disponible y asigna el primero
				for (Especialidad tipoProblema : paciente.getTipoProblemasMedicos()) {
					cirujano = this.getcirujanoEspecialidad(tipoProblema);
					if(cirujano != null) { break; }
				};
				
				// Si no hay doctores disponibles, entonces conseguir un cirujano cualquiera
//				if(cirujano == null) { 
//					cirujano = this.getcirujanoEspecialidad(Especialidad.GENERAL); 
//				}
				
				// Si no hay cirujanos disponibles, entonces conseguir un doctor general para que opere
				// for(Especialidad)
				
				// Si definitiva no encontro doctor, entonces F
				if(cirujano == null) { return null; }
				
				// Si llego hasta aca es que si hay cirujano, por lo que creemos uno
				// para que ese elemento de la lista no apunte hacia un null sino hacia un Quirofano
				quirofanos[i] = new Quirofano(paciente, cirujano);
				return quirofanos[i];
			}
		}
		// Si se retorno null es porque no hay quirofanos disponibles
		return null;
	}
	
	/*
	 * getDoctorEspecialidad(Especialidad especialidad)
	 */
	public Doctor getcirujanoEspecialidad(Especialidad especialidad) {
		for (Doctor doctor : this.getDoctores()) {
				if(doctor != null && doctor.getEspecialidad() == especialidad && doctor.isCirujano()) {
				return doctor;
			}
		}
		return null;
	}

	/*
	 * esparcirVirus: se esparce el virus a los pacientes en la sala de espera
	 * puede ser el caso que el paciente se muera al ser poblacion de riesgo y por mala suerte(azar)
	 */
	public ArrayList<Persona> esparcirVirus() {
		ArrayList<Persona> personasHospital = getPersonasHospital();
		ArrayList<Persona> muertos = new ArrayList<Persona>();
		Random rand = new Random();
		
		for (Persona persona : personasHospital) {
			if(persona.isPoblacionDeRiesgo() && rand.nextBoolean()) {
				boolean muerto = persona.infectar();
				if(muerto) {
					muertos.add(persona);
					if(persona instanceof Paciente) {
						this.getSalaDeEspera().sacarPaciente(persona.getCedula());						
					} else if(persona instanceof Doctor) {
						this.despedirDoctor(persona.getCedula(), true);
					}
				}
			}
		}
		return muertos;
	}

	
	
	/*
	 * 
	 */
	public ArrayList<Persona> getPersonasHospital() {
		ArrayList<Persona> personasHospital = new ArrayList<Persona>();
		
		for(Persona persona : getSalaDeEspera().getPacientes()) {
			if(persona != null) {
				personasHospital.add(persona);
			}
		}
		
		for(Persona persona : getDoctores()) {
			if(persona != null) {
				personasHospital.add(persona);
			}
		}
		return personasHospital;
	}
	
	/*	GETTERS & SETTERS	*/
	public ArrayList<Cita> getHorario() { return this.horario; }
	public void delCita(Cita cita) {  horario.remove(cita); }
	public void addCita(Cita cita) { horario.add(cita); }
	
	public int getDenuncias() { return this.denuncias; }
	public void setDenuncias(int denuncias) { this.denuncias = denuncias; }
	
	public SalaDeEspera getSalaDeEspera() { return this.salaDeEspera; }
	public void setSalaDeEspera(SalaDeEspera salaDeEspera) { this.salaDeEspera = salaDeEspera; }
	
	public Quirofano[] getQuirofanos() { return this.quirofanos; }
	public void setQuirofanos(Quirofano[] quirofanos) { this.quirofanos = quirofanos;}
	
	public Cuarto[] getCuartos() { return this.cuartos; }
	public void setCuartos(Cuarto[] cuartos) { this.cuartos = cuartos; }
	
	public Doctor[] getDoctores() { return this.doctores; }
	public void setDoctores(Doctor[] doctores) { this.doctores = doctores; }
	
	public int getPresupuesto() { return this.presupuesto; }
	public void setPresupuesto(int presupuesto) { this.presupuesto = presupuesto; }

	public boolean isBioseguro() { return this.bioseguro; }
	public void setBioseguro(boolean bioseguro) { this.bioseguro = bioseguro; }
	
}
